src/main/java/bio/terra/workspace/service/workspace/flight/GoogleCloudSyncStep.java

```diff
@@ -0,0 +1,103 @@
+ * changes. The read-modify-write pattern is necessary to support existing IAM groups which the
+ * Buffer Service may grant on projects before handing them out.
+ *
+ * <p>The "modify" part of this step specifically adds GCP bindings as specified in {@link
+ * CloudSyncRoleMapping}. Note that the bindings list sent to GCP may contain multiple entries with
+ * the same role. This is valid, though GCP will condense them into one binding per role internally.
+ */
+public class GoogleCloudSyncStep implements Step {
+
+  CloudResourceManagerCow resourceManagerCow;
+
+  private Logger logger = LoggerFactory.getLogger(GoogleCloudSyncStep.class);
+
+  @Autowired
+  public GoogleCloudSyncStep(CloudResourceManagerCow resourceManagerCow) {
+    this.resourceManagerCow = resourceManagerCow;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    String gcpProjectName = flightContext.getWorkingMap().get(GOOGLE_PROJECT_ID, String.class);
+    FlightMap workingMap = flightContext.getWorkingMap();
+    try {
+      Policy currentPolicy =
+          resourceManagerCow
+              .projects()
+              .getIamPolicy(gcpProjectName, new GetIamPolicyRequest())
+              .execute();
+      List<Binding> existingBindings = currentPolicy.getBindings();
+
+      Map<IamRole, String> samEmailMap =
+          Map.of(
+              IamRole.READER,
+                  workingMap.get(WorkspaceFlightMapKeys.IAM_READER_GROUP_EMAIL, String.class),
+              IamRole.WRITER,
+                  workingMap.get(WorkspaceFlightMapKeys.IAM_WRITER_GROUP_EMAIL, String.class),
+              IamRole.OWNER,
+                  workingMap.get(WorkspaceFlightMapKeys.IAM_OWNER_GROUP_EMAIL, String.class));
+      List<Binding> newBindings = new ArrayList<>();
+      for (IamRole iamRole : IamRole.values()) {
+        // GCP IAM always prefixes groups with the literal "group:"
+        String groupEmail = "group:" + samEmailMap.get(iamRole);
+        for (String gcpRole : CloudSyncRoleMapping.cloudSyncRoleMap.get(iamRole)) {
+          newBindings.add(
+              new Binding().setRole(gcpRole).setMembers(Collections.singletonList(groupEmail)));
+        }
+      }
+      // Add all existing bindings to ensure we don't accidentally clobber existing permissions.
+      newBindings.addAll(existingBindings);
+      Policy newPolicy =
+          new Policy()
+              .setVersion(currentPolicy.getVersion())
+              .setBindings(newBindings)
+              .setEtag(currentPolicy.getEtag());
+      SetIamPolicyRequest iamPolicyRequest = new SetIamPolicyRequest().setPolicy(newPolicy);
+      logger.info("Setting new Cloud Context IAM policy: " + iamPolicyRequest.toPrettyString());
+      resourceManagerCow.projects().setIamPolicy(gcpProjectName, iamPolicyRequest).execute();
+    } catch (IOException e) {
+      throw new RetryException("Error setting IAM permissions", e);
```

Comment:
Should this be a more specific exception? I don't understand the pattern for making new exceptions in WM.

Reply:
In general I try to create a specific exception when it would provide some useful context for debugging or logs. This could reasonably be something like `RetryableCrlException`, so I added that for this + other steps.